### PR TITLE
Allow to define prop type using native types constructors

### PR DESCRIPTION
**Pull request by blikblum**
* [ ] Bug
* [x] Feature

## Requirements

* [x] Read the [contribution guidelines](https://github.com/skatejs/skatejs/blob/5.x/CONTRIBUTING.md).
* [x] Wrote tests.
* [x] Updated docs and upgrade instructions, if necessary.

## Rationale

Currently is necessary to import props from skatejs package to define simple props types like string, number etc

Given the skatejs design based on mixins, most of the times, this is the only import from skatejs when defining a component class

By allowing to use native constructors like String, Number the props import could be avoided. It's also pretty intuitive and align with other libraries like [lit-element](https://github.com/Polymer/lit-element/)

With this change is possible to do

```
class Test extends MyComponent {
  static props = {
     str: String,
     obj: Object
  }
}
```

## Implementation

Used a Map instance to store the props definitions mapped to correspondent native constructor

## Open questions

Should props.any be mapped to null?
Should be added a function `registerPropType` to allow  sharing custom properties? 

## Other

To test if props.* match to constructors needed to access private field _propsNormalised in tests
In normaliseAttributeDefinition i created a identity function to be used when necessary instead of creating a anonymous function on the fly  

## Tasks

_List any tasks you need to do here, if any. Delete this section if you don't need it._

* [ ] Document

### CODE REVIEW

**Comment by treshugart** [comment]
> Should props.any be mapped to null?

Yeah, that's probably a good idea.

> Should be added a function registerPropType to allow sharing custom properties?

How would this work?

**Comment by blikblum** [comment]
I implemented null/undefined map to props.any using it as a fallback in defineProps:
`let func = props[name] || props.any;`

Let me know if you prefer to do the mapping through defaultTypes instance


>> Should be added a function registerPropType to allow sharing custom properties?

> How would this work?

```javascript
// implementation
registerPropType (typeId, definition) {
  defaultTypes.set(typeId, definition)
}

// setup
import { registerPropType } from 'skatejs'

registerPropType('mytype', {...TypeDef})

//usage
class Comp extends Component {
  static props = {
     myProp: 'mytype'
  }
}
```
I think if aggred to implemented, this should be done in a separated PR

File: packages/skatejs/test/unit/with-update.spec.js [hunk 1]
```diff
@@ -98,6 +98,9 @@ test('prop types could be defined with native types constructors', () => {
   class Test extends withUpdate() {
     static get props() {
       return {
+        propsAny: props.any,
```
Comment (treshugart) [inline packages/skatejs/test/unit/with-update.spec.js]: I'm actually fine if we assert something like:

```js
expect(Component.props.nativeNull).toMatchObject(props.any);
```

And remove direct testing of `props.any` in withUpdate (and of course continue testing them separately as a unit).

Comment (treshugart) [reply packages/skatejs/test/unit/with-update.spec.js]: I'd also like to start advocating the usage of `null`, `String`, `Array`, `Object`, etc over the `props.*` as it's definitely closer to the metal in how one might expect it to behave. Custom usage would involve the `props.*` API.

Comment (treshugart) [reply packages/skatejs/test/unit/with-update.spec.js]: This might be out of the scope of this PR and we can do it in 6, though. For now, I think I'm happy to see this as is, but your thoughts here would be useful.

**Comment by treshugart** [comment]
I'm not sure if we need that API. I think we should be covering all built-ins with pre-configured prop types (like we are). Any edge cases, consumers can just write custom prop definitions, as they can now. I'm generally not a fan of having a function that mutates the global configuration of a library.

File: packages/skatejs/test/unit/with-update.spec.js (hunk 1 above)
Comment (blikblum) [reply packages/skatejs/test/unit/with-update.spec.js]: > expect(Component.props.nativeNull).toMatchObject(props.any);

As far as i remember this did not work because of the prop name saved in prop definition. I can try again and put the error here

> And remove direct testing of props.any in withUpdate (and of course continue testing them separately as a unit).

Do you mean creating a separated spec file?

> This might be out of the scope of this PR and we can do it in 6, though. For now, I think I'm happy to see this as is, but your thoughts here would be useful.

Agree

Comment (treshugart) [reply packages/skatejs/test/unit/with-update.spec.js]: I'm happy to keep as-is for the short-term.

**Comment by treshugart** [comment]
What work is left here? It still looks like you need to implement the mapping between native / `props.*`?

**Comment by blikblum** [comment]
> What work is left here? It still looks like you need to implement the mapping between native / props.*?

The mapping (Array -> props.array etc) is already done. 

Needs to update the docs

**Comment by blikblum** [comment]
Updated docs

**Comment by treshugart** [comment]
> Also there's a misspelling normalised -> normalized but initially i think better todo in a separated PR

Oh, ha! That's "proper" English (UK / AU). I moved from California to Sydney, so I switch between the two without knowing it.

File: README.md [hunk 2]
```diff
@@ -60,7 +60,7 @@ import { h } from 'preact';
 class WithPreact extends withComponent(withPreact()) {
   static get props() {
     return {
-      name: props.string
+      name: props.string // String could be used also to define the prop type
```
Comment (treshugart) [inline README.md]: I'd actually prefer we advocate native usage instead of built-in, and introduce custom usage separately. We can do that in a separate PR.

**Comment by treshugart** [comment]
This is awesome! Last thing to do is to pull latest and run `bolt` to ensure deps are latest. That should green the build.

**Comment by blikblum** [comment]
I updated the repo, rebased and run yarn / bolt but the skatejs dependency on site package was not updated automatically. I had to update it manually.

With this the tests fail at MutationObserver tests. I checked and found that the errors were introduced by  https://github.com/skatejs/skatejs/commit/932bd95a9d141dce30b6324df573b91e32c10c74

**Comment by treshugart** [comment]
Okay, I'll see what's going on there.

**Comment by blikblum** [comment]
I rebased and now CI/Travis is passing

**Comment by treshugart** [comment]
Nice work! Thanks!

**Comment by treshugart** [comment]
Released in 5.2.0.
